src/nfa/nfa.cc

```diff
@@ -193,21 +193,155 @@ void Nfa::add_trans(State stateFrom, Symbol symbolOnTransition, State stateTo) {
     auto transitionFromStateIter = transitionrelation[stateFrom].begin();
     for (; transitionFromStateIter != transitionrelation[stateFrom].end(); ++transitionFromStateIter) {
         if (transitionFromStateIter->symbol == symbolOnTransition) {
+            // Add transition with symbolOnTransition already used on transitions from stateFrom.
             transitionFromStateIter->states_to.insert(stateTo);
             return;
         } else if (transitionFromStateIter->symbol > symbolOnTransition) {
             break;
         }
     }
 
+    // Add transition to a new TransSymbolStates struct with symbolOnTransition yet unused on transitions from stateFrom.
     transitionrelation[stateFrom].insert(transitionFromStateIter, TransSymbolStates(symbolOnTransition, stateTo));
 }
 
+void Nfa::remove_trans(State src, Symbol symb, State tgt)
+{
+    if (!has_trans(src, symb, tgt))
+    {
+        throw std::invalid_argument(
+                "Transition [" + std::to_string(src) + ", " + std::to_string(symb) + ", " +
+                std::to_string(tgt) + "] does not exist.");
+    }
+
+    auto transitionFromStateIter{ transitionrelation[src].begin() };
+    while (transitionFromStateIter != transitionrelation[src].end())
+    {
+        if (transitionFromStateIter->symbol == symb)
+        {
+            transitionFromStateIter->states_to.remove(tgt);
+            if (transitionFromStateIter->states_to.empty())
+            {
+                transitionrelation[src].remove(*transitionFromStateIter);
+            }
+
+            return;
+        }
+
+        ++transitionFromStateIter;
+    }
+}
+
 State Nfa::add_new_state() {
     transitionrelation.emplace_back();
     return transitionrelation.size() - 1;
 }
 
+void Nfa::remove_epsilon(const Symbol epsilon)
+{
+    *this = Mata::Nfa::remove_epsilon(*this, epsilon);
+}
+
+StateSet Nfa::get_reachable_states() const
+{
+    StateBoolArray reachable_bool_array{ compute_reachability() };
+
+    StateSet reachable_states{};
+    for (State original_state{ 0 }; original_state < get_num_of_states(); ++original_state)
```

Comment:
I wonder, whether this could be done more efficiently, since you basically have 2n algorithm (first you count reachability in `n` steps, and then you iterate the states again in `n` steps). Seems to be potential for optimisation.

Reply:
It could be. I will investigate. Thank you for pointing that out.

Reply:
Correct me where/if I am wrong. I see this as follows:

I perform a reachability computation by starting in initial states and following transitions to other states. The complexity of this operation is `n` where `n` is the number of reachable states in the automaton (I will not iterate over unreachable states). Then, I iterate over all states in the automaton, adding only the reachable states to the newly created state set. Here, I make use of the underlaying structure: vector. The states can be simply pushed to the back of the vector (because they are already sorted thanks to the `for` loop). `std::vector::push_back()` has amortized constant complexity.

Therefore, we have `n + n'` time complexity (`n` is the number of reachable states, `n'` is the number of all states).

However, if I were to rewrite the algorithm in such a way that I would iterate over automata states only once (insert a new state to the state set each time I encounter a new reachable state in `compute_reachability()`), I would need to perform a binary search for each insertion (as implemented by `Ordvector::insert()`). I cannot be sure that the newly discovered reachable state has the largest number of all reachable states found so far. Hence, I cannot push to the back of the `OrdVector` and a binary search is necessary. That would create `n * log n` time complexity as, in the worst case scenario, each state could be reachable.

I believe the current implementation is more optimized than the above proposed alternative. This boils down
to the advantages and disadvantages of sorted vector as a state set. I could return a vector of unsorted states instead. However, then I could not use `Nfa::StateSet` or I would need to sort the vector of unsorted reachable states in the end and allocate a new `StateSet` with the sorted vector. That would generate `n' + n` time complexity again (or even more if we assert the `StateSet` is sorted for each insert), if I am not mistaken.

Reply:
Yes, that approach is definitely less effective. I hoped, that the `compute_reachability` could work with `vector<State>` instead of boolean. But you are right, that you would either had to use the logarithmic insert to OrdVector or sort the vector of reachable states, which would not save us much computation time. Consider this suggestion thus as resolved.

Reply:
> sort the vector of reachable states afterwards to create OrdVector, which would not save us much computation time

Due to the implementation of `Ordvector(vec)` constructor, I believe it would be even more expensive.

Furthermore, we make use of `std::vector<bool>` for its space-efficient implementation (suggested by C++ standard), which should result in a better performance when allocating these vectors for large automata than creating `std::vector<State> reachable_states` and then calling `StateSet(reachable_states)` (which is `OrdVector<State>`).